From 138f81bd8eb52c02e15e5d18ea48903b69219d46 Mon Sep 17 00:00:00 2001
From: Oliver Eftevaag <oliver.eftevaag@qt.io>
Date: Fri, 19 Sep 2025 16:38:48 +0200
Subject: Update controls theme on QGuiApp::styleHints palette or theme change

Previously, we would only update the theme for a style when
QGuiApplication::styleHints()->colorSchemeChanged() would be emitted.
This was unreliable in cases where the user changes the system theme to
a different theme that uses different colors, but the same color scheme.
In that case, the theme wouldn't update, resulting in stale palette
colors.

With 8e7b97bba513d3ba5526a1827bbda3470dd7d02c now forwarding theming
related events to QGuiApp::styleHints, we can install an event filter on
it, and react to both application palette changes and theme changes.

Fixes: QTBUG-140507
Pick-to: 6.11 6.10 6.8
Change-Id: I14df2b7c82a843e7a40cc17f587fc0de3c37c372
Reviewed-by: Volker Hilsheimer <volker.hilsheimer@qt.io>
(cherry picked from commit 878f9ee744e5dc360ba419217851f42a2f52a8f2)
---
 src/quickcontrols/qquickstyleplugin.cpp       | 39 +++++++-
 src/quickcontrols/qquickstyleplugin_p.h       |  5 +
 .../qquickmaterialstyle/data/tst_material.qml |  2 +
 .../tst_qquickmaterialstyle.cpp               |  1 +
 .../qquickstyle/tst_qquickstyle.cpp           | 95 +++++++++++++++++++
 5 files changed, 138 insertions(+), 4 deletions(-)

diff --git x/qtdeclarative/src/quickcontrols/qquickstyleplugin.cpp y/qtdeclarative/src/quickcontrols/qquickstyleplugin.cpp
index 5e94843bc5096697e962ef4de73a1202816fc620..372d5f39a101df8215eb911a69d49e0756b8f571 100644
--- x/qtdeclarative/src/quickcontrols/qquickstyleplugin.cpp
+++ y/qtdeclarative/src/quickcontrols/qquickstyleplugin.cpp
@@ -19,6 +19,35 @@ QT_BEGIN_NAMESPACE
 
 Q_STATIC_LOGGING_CATEGORY(lcStylePlugin, "qt.quick.controls.styleplugin")
 
+class QQuickThemeChangeObserver : public QObject {
+    Q_OBJECT
+public:
+    explicit QQuickThemeChangeObserver()
+    {
+        QStyleHints *styleHints = QGuiApplication::styleHints();
+        moveToThread(styleHints->thread());
+        styleHints->installEventFilter(this);
+    }
+
+Q_SIGNALS:
+    void paletteOrThemeChanged();
+
+protected:
+    bool eventFilter(QObject *object, QEvent *event) override {
+        Q_UNUSED(object);
+        if (event->type() == QEvent::ApplicationPaletteChange || event->type() == QEvent::ThemeChange)
+            emit paletteOrThemeChanged();
+        return false;
+    }
+};
+
+// The custom destructor ensures that all posted events gets processed before the event filter is deleted.
+void QQuickStylePlugin::ObserverDeleter::operator()(QQuickThemeChangeObserver *observer)
+{
+    observer->disconnect();
+    observer->deleteLater();
+}
+
 QQuickStylePlugin::QQuickStylePlugin(QObject *parent)
     : QQmlExtensionPlugin(parent)
 {
@@ -76,8 +105,10 @@ void QQuickStylePlugin::registerTypes(const char *uri)
             << QQuickStylePrivate::fallbackStyle() << "; calling initializeTheme()";
     }
     initializeTheme(theme);
-    connect(QGuiApplication::styleHints(), &QStyleHints::colorSchemeChanged,
-                                     this, &QQuickStylePlugin::updateTheme);
+
+    themeChangeObserver.reset(new QQuickThemeChangeObserver);
+    connect(themeChangeObserver.get(), &QQuickThemeChangeObserver::paletteOrThemeChanged,
+            this, &QQuickStylePlugin::updateTheme);
 
     if (!isPrimaryFallback && !styleName.isEmpty())
         QFileSelectorPrivate::addStatics(QStringList() << styleName);
@@ -89,8 +120,7 @@ void QQuickStylePlugin::unregisterTypes()
     if (!QQuickThemePrivate::instance)
         return;
 
-    disconnect(QGuiApplication::styleHints(), &QStyleHints::colorSchemeChanged,
-                                        this, &QQuickStylePlugin::updateTheme);
+    themeChangeObserver.reset();
 
     // Not every style has a plugin - some styles are QML-only. So, we clean this
     // stuff up when the first style plugin is unregistered rather than when the
@@ -138,4 +168,5 @@ QQuickTheme *QQuickStylePlugin::createTheme(const QString &name)
 
 QT_END_NAMESPACE
 
+#include "qquickstyleplugin.moc"
 #include "moc_qquickstyleplugin_p.cpp"
diff --git x/qtdeclarative/src/quickcontrols/qquickstyleplugin_p.h y/qtdeclarative/src/quickcontrols/qquickstyleplugin_p.h
index 726af7c5ed59c541532def1c7d339b933424d2b7..332a06ef55a2dd2a11ef49f8b06b221252c7820f 100644
--- x/qtdeclarative/src/quickcontrols/qquickstyleplugin_p.h
+++ y/qtdeclarative/src/quickcontrols/qquickstyleplugin_p.h
@@ -18,10 +18,12 @@
 
 #include <QtQml/qqmlextensionplugin.h>
 #include <QtQuickControls2/qtquickcontrols2global.h>
+#include <memory>
 
 QT_BEGIN_NAMESPACE
 
 class QQuickTheme;
+class QQuickThemeChangeObserver;
 
 class Q_QUICKCONTROLS2_EXPORT QQuickStylePlugin : public QQmlExtensionPlugin
 {
@@ -41,6 +43,9 @@ public:
 private:
     QQuickTheme *createTheme(const QString &name);
 
+    struct ObserverDeleter { void operator()(QQuickThemeChangeObserver *observer); };
+    std::unique_ptr<QQuickThemeChangeObserver, ObserverDeleter> themeChangeObserver;
+
     Q_DISABLE_COPY(QQuickStylePlugin)
 };
 
diff --git x/qtdeclarative/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml y/qtdeclarative/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml
index 47ed0b95a040e25fe903b3552eb9439d34866d97..4545a783043d4824a47b49ae9420ec830e068bbe 100644
--- x/qtdeclarative/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml
+++ y/qtdeclarative/tests/auto/quickcontrols/qquickmaterialstyle/data/tst_material.qml
@@ -1574,7 +1574,9 @@ TestCase {
         let window = createTemporaryObject(systemThemeComponent, testCase)
         verify(window)
 
+        // The theme might be Material.System initially, but that's fine, since we'll just set it to Dark immediately
         const toggleTheme = (theme) => (theme === Material.Dark) ? Material.Light : Material.Dark
+        verify(TestHelper.platformTheme !== undefined)
 
         TestHelper.platformTheme = toggleTheme(TestHelper.platformTheme)
         tryCompare(window.Material, "theme", TestHelper.platformTheme)
diff --git x/qtdeclarative/tests/auto/quickcontrols/qquickmaterialstyle/tst_qquickmaterialstyle.cpp y/qtdeclarative/tests/auto/quickcontrols/qquickmaterialstyle/tst_qquickmaterialstyle.cpp
index 56e27a27b58c28d9ab9664eb633b6e9b5391234a..15e90ab03ca5060d7dc2946a8573512b6cf743e1 100644
--- x/qtdeclarative/tests/auto/quickcontrols/qquickmaterialstyle/tst_qquickmaterialstyle.cpp
+++ y/qtdeclarative/tests/auto/quickcontrols/qquickmaterialstyle/tst_qquickmaterialstyle.cpp
@@ -41,6 +41,7 @@ class Setup : public QObject
 {
     Q_OBJECT
     Q_PROPERTY(QQuickMaterialStyle::Theme platformTheme READ platformTheme WRITE setPlatformTheme FINAL)
+    QML_EXTENDED_NAMESPACE(QQuickMaterialStyle)
 
 public slots:
     void applicationAvailable()
diff --git x/qtdeclarative/tests/auto/quickcontrols/qquickstyle/tst_qquickstyle.cpp y/qtdeclarative/tests/auto/quickcontrols/qquickstyle/tst_qquickstyle.cpp
index 8806b04a1e37e2aeaa55e9ea9491df35aebad1a8..388faee6467972e5a45f745d474812fab13008de 100644
--- x/qtdeclarative/tests/auto/quickcontrols/qquickstyle/tst_qquickstyle.cpp
+++ y/qtdeclarative/tests/auto/quickcontrols/qquickstyle/tst_qquickstyle.cpp
@@ -11,6 +11,7 @@
 #include <QtQuickTemplates2/private/qquicklabel_p_p.h>
 #include <QtQuickTemplates2/private/qquicktheme_p.h>
 #include <QtGui/private/qguiapplication_p.h>
+#include <QtGui/qpa/qplatformtheme.h>
 
 class tst_QQuickStyle : public QQmlDataTest
 {
@@ -26,6 +27,8 @@ private slots:
     void configurationFile();
     void commandLineArgument();
     void environmentVariables();
+    void qGuiApplicationPaletteChangesArePropagatedToControls();
+    void defaultPaletteIsUpdatedWhenChangingColorScheme();
 
 private:
     Q_REQUIRED_RESULT bool loadControls();
@@ -143,6 +146,98 @@ void tst_QQuickStyle::environmentVariables()
     QCOMPARE(QQuickStylePrivate::fallbackStyle(), QString());
 }
 
+void tst_QQuickStyle::qGuiApplicationPaletteChangesArePropagatedToControls()
+{
+    QVERIFY(!QQuickTheme::instance());
+    QQuickStyle::setStyle("Fusion");
+    QCOMPARE(QQuickStyle::name(), QStringLiteral("Fusion"));
+    QQmlEngine engine;
+    engine.addImportPath(dataDirectory());
+    QQmlComponent component(&engine);
+    component.setData("import QtQuick; import QtQuick.Controls; ApplicationWindow { Label { anchors.centerIn: parent; } }", QUrl());
+
+    QScopedPointer<QObject> object(component.create());
+    QVERIFY(!object.isNull());
+    auto *label = object->findChild<QQuickLabel *>();
+    QVERIFY(label);
+    auto *labelPrivate = QQuickLabelPrivate::get(label);
+
+    // The fusion style causes QQuickTheme::palette() to use the platform themes palette directly.
+    QVERIFY(QQuickTheme::instance()->usePlatformPalette());
+
+    QPalette qGuiAppPalette = QGuiApplication::palette();
+    const QPalette* qPlatformThemePalette = QGuiApplicationPrivate::platformTheme()->palette(QPlatformTheme::LabelPalette);
+
+    // The initial QGuiApplication::palette() is usually built from the platform theme's system palette.
+    // The platform theme can have multiple other "role" palettes, including a LabelPalette, which is what QQuickTheme will query for.
+    // Since the initial QGuiApplication::palette() typically has a resolveMask of 0, it will not be prioritized over the defaultPalette.
+    // Hence we expect the label's palette and default palette to both be equal to the platform theme's LabelPalette,
+    // which differs from the system palette on macOS. However, if the platform theme returns nullptr for LabelPalette (which happens when the platform theme is "offscreen"),
+    // we fallback to a gray palette that is in sync with QGuiApplication::palette().
+    const QColor expectedInitialWindowTextColor = qPlatformThemePalette ? qPlatformThemePalette->windowText().color() : qGuiAppPalette.windowText().color();
+    QCOMPARE(labelPrivate->palette()->windowText(), expectedInitialWindowTextColor);
+    QCOMPARE(labelPrivate->defaultPalette().windowText().color(), expectedInitialWindowTextColor);
+
+    qGuiAppPalette.setColor(QPalette::WindowText, QColorConstants::Magenta);
+    QGuiApplication::setPalette(qGuiAppPalette);
+
+    QTRY_COMPARE(labelPrivate->palette()->active()->windowText(), QColorConstants::Magenta);
+    QCOMPARE(qGuiAppPalette.windowText().color(), QColorConstants::Magenta);
+    // When the QQuickTheme uses the platform palette, QQuickTheme::palette(Scope scope) will take the palette directly from
+    // QGuiApplicationPrivate::platformTheme(), skipping any potential changes made to QGuiApplication::palette().
+    // We'll thus have a mismatch between the Label's palette (which is updated and propagated from QQuickWindow::palette)
+    // and default palette (which is taken from the platform theme's palette).
+    QCOMPARE_NE(labelPrivate->defaultPalette().windowText().color(), QColorConstants::Magenta);
+    QCOMPARE(labelPrivate->defaultPalette().windowText().color(), expectedInitialWindowTextColor);
+}
+
+void tst_QQuickStyle::defaultPaletteIsUpdatedWhenChangingColorScheme()
+{
+    QVERIFY(!bool(QQuickTheme::instance()));
+    QQuickStyle::setStyle("Basic");
+    QCOMPARE(QQuickStyle::name(), QStringLiteral("Basic"));
+    QQmlEngine engine;
+    engine.addImportPath(dataDirectory());
+    QQmlComponent component(&engine);
+    component.setData("import QtQuick; import QtQuick.Controls; Label { }", QUrl());
+
+    QScopedPointer<QObject> object(component.create());
+    QVERIFY(!object.isNull());
+
+    auto *labelPrivate = QQuickLabelPrivate::get(qobject_cast<QQuickLabel *>(object.data()));
+    QColor oldLabelWindowText = labelPrivate->defaultPalette().windowText().color();
+    QColor oldLabelWindow = labelPrivate->defaultPalette().window().color();
+
+    QPalette oldSystemPalette = QQuickTheme::palette(QQuickTheme::System);
+
+    auto oldColorScheme = QGuiApplication::styleHints()->colorScheme();
+    auto guard = qScopeGuard([](){
+        QGuiApplication::styleHints()->unsetColorScheme();
+    });
+
+    if (oldColorScheme == Qt::ColorScheme::Dark)
+        QGuiApplication::styleHints()->setColorScheme(Qt::ColorScheme::Light);
+    else if (oldColorScheme == Qt::ColorScheme::Light)
+        QGuiApplication::styleHints()->setColorScheme(Qt::ColorScheme::Dark);
+    else
+        QSKIP(qPrintable(QLatin1String("Skipping test. "
+            "The %1 platform theme likely lacks a proper colorScheme() and requestColorScheme(Qt::ColorScheme) implementation.").arg(qGuiApp->platformName())));
+
+    // Colors should be updated
+    QTRY_COMPARE_NE(QQuickTheme::palette(QQuickTheme::System), oldSystemPalette);
+
+    QCOMPARE_NE(labelPrivate->defaultPalette().windowText().color(), oldLabelWindowText);
+    QCOMPARE_NE(labelPrivate->defaultPalette().window().color(), oldLabelWindow);
+
+    if (QGuiApplication::styleHints()->colorScheme() == Qt::ColorScheme::Light) {
+        QCOMPARE_GT(labelPrivate->defaultPalette().window().color().lightnessF(), 0.5);
+        QCOMPARE_LT(labelPrivate->defaultPalette().windowText().color().lightnessF(), 0.5);
+    } else if (QGuiApplication::styleHints()->colorScheme() == Qt::ColorScheme::Dark) {
+        QCOMPARE_LT(labelPrivate->defaultPalette().window().color().lightnessF(), 0.5);
+        QCOMPARE_GT(labelPrivate->defaultPalette().windowText().color().lightnessF(), 0.5);
+    }
+}
+
 QTEST_MAIN(tst_QQuickStyle)
 
 #include "tst_qquickstyle.moc"
